### PR TITLE
Adapt the contract to simplify UBQ farming #243

### LINKED ISSUE #243: Adapt the contract to simplify UBQ farming
### Background 

The source code in this repo https://github.com/ubiquity/ubiquity-instadapp/tree/master/backend/contracts/mainnet 

We made this contract for Instadapp 
https://etherscan.io/address/0x8ec066d75d665616a94f2eccdbe49b54eaeefc78#code

### To do 
Adapt the contract to simplify UBQ farming

- Adapt the contract for Instadapp for the Ubiquity website to simplify liquidity providing for DeFi amateurs. The first step of this assignment is to adapt zapaz's code (remove all the instadapp specific code) to allow users to directly farm from our UI.

- The second part will be eventually to make a UI component to operate this smart contract.

### Acceptance Criteria 
TBD

### PR DESCRIPTION
**Pull request by kfujita0520**
This is the PR for #243.
To reflect feedback from my initial draft commit, I have removed hardcode contract address.
https://github.com/ubiquity/ubiquity-dollar/issues/243#issuecomment-1267956556

The variable name is generalized, but still assume we are using MetaPool liquidity with Curve 3Pool base and uAD. 
(Curve pool can change the 3 base token, you could change bonding address)
If this base assumption has changed, we should rewrite the logic or contract again. (I can help as separate task at that time.)

Foundry test case did not work as I have escalated in telegram chat group. 
However, I have tested with java script with hardhat on https://github.com/kfujita0520/simplify-ubq-farming
The foundry test script where i have faced a challenge is in this repository.
https://github.com/kfujita0520/Foundry_Test
I have ask a question in Foundry community but could not get any response.

### CODE REVIEW

**Comment by 0x4007** [comment]
It looks like you updated a lot of interfaces. Is this necessary? We would prefer to only change what is necessary. 

I requested a review from @0xcodercrane because he is a Solidity developer and he knows our protocol very well. It isn't clear to me how I can test this contract either. Keep me posted @0xcodercrane

File: packages/contracts/src/dollar/interfaces/IERC1155Ubiquity.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IERC1155Ubiquity.sol:25]: Curious why this was added too.

**Review by 0x4007 — CHANGES_REQUESTED** [review]

**Comment by cuddle-dev** [comment]
> It looks like you updated a lot of interfaces. Is this necessary? We would prefer to only change what is necessary.
> 
> I requested a review from @0xcodercrane because he is a Solidity developer and he knows our protocol very well. It isn't clear to me how I can test this contract either. Keep me posted @0xcodercrane

I will take a deep look and keep you updated

**Comment by kfujita0520** [comment]
> It looks like you updated a lot of interfaces. Is this necessary?

@pavlovcik Thank you very much for your comments and review. 
The main logic resides in only one contract: DirectUBQFarmer(DirectGovernanceFarmer). 
the rest of contract is just interfaces which need to interact with external contract and nothing impact on your existing system.
Meanwhile I changed one existing interface
```
-    function holderTokens() external view returns (uint256[] memory);
+    function holderTokens(address holder) external view returns (uint256[] memory);
```

I believe this was the bug of your interface code and I have fixed it. Because your implementation contract has address argument on holderTokens method and that is inconsistent with your original interface. For your information, here is the link to your  BondingShareV2 contract on mainnet.
https://etherscan.io/address/0x2dA07859613C14F6f05c97eFE37B9B4F212b5eF5#readContract

Thank you very much.

File: packages/contracts/src/dollar/interfaces/IERC1155Ubiquity.sol
Comment (cuddle-dev) [reply packages/contracts/src/dollar/interfaces/IERC1155Ubiquity.sol:25]: here @kfujita0520  is right. we were missing `address holder` parameter.

**Review by cuddle-dev — CHANGES_REQUESTED** [review]
codebase looks good. We want to have unit tests in forge and take another look once unit tests done before merging.  You can refer a bunch of unit tests from ubiquity:development branch if you have some environment setup issue. 
I also want @zgorizzo69 's review for this pull request if you have time, I don't have much context about instadapp

**Comment by 0x4007** [comment]
It was actually @zapaz that worked on Instadapp. The purpose of it primarily was to integrate Ubiquity into the instadapp dashboard. That way we would have more eyes on our project. We had to conform to their standards to do so. 

After setting up the scaffolding (conforming to their standard) we wanted to allow people to easily "ape in" and "ape out" of our staking. Meaning that if they have DAI/USDC/USDT/uAD in their wallet, they can press a single button to:

1. Deposit that liquidity in our metapool 
2. Stake the LP token

As well as the opposite. 

The purpose of "ape in" and "ape out" was to reduce the friction of adding to our dollar's liquidity. 

As I understand, the logic that I describe above is all implemented already in the instadapp smart contracts that @zapaz wrote. 

Now we should extract that logic, and create new clean contracts (without the instadapp scaffolding) and allow users to "ape in" and "ape out" from our dashboard.

File: packages/contracts/src/dollar/interfaces/IERC1155Ubiquity.sol
Comment (0x4007) [reply packages/contracts/src/dollar/interfaces/IERC1155Ubiquity.sol:25]: > > It looks like you updated a lot of interfaces. Is this necessary?
> 
> @pavlovcik Thank you very much for your comments and review. The main logic resides in only one contract: DirectUBQFarmer(DirectGovernanceFarmer). the rest of contract is just interfaces which need to interact with external contract and nothing impact on your existing system. Meanwhile I changed one existing interface
> 
> ```
> -    function holderTokens() external view returns (uint256[] memory);
> +    function holderTokens(address holder) external view returns (uint256[] memory);
> ```
> 
> I believe this was the bug of your interface code and I have fixed it. Because your implementation contract has address argument on holderTokens method and that is inconsistent with your original interface. For your information, here is the link to your BondingShareV2 contract on mainnet. https://etherscan.io/address/0x2dA07859613C14F6f05c97eFE37B9B4F212b5eF5#readContract
> 
> Thank you very much.

I appreciate the attention to detail @kfujita0520. Great job!

**Comment by 0x4007** [comment]
Hey @kfujita0520 just following up on this- is everything all right on your end?

**Comment by kfujita0520** [comment]
@0xcodercrane 
>We want to have unit tests in forge and take another look once unit tests done before merging. You can refer a bunch of unit tests from ubiquity:development branch if you have some environment setup issue.

Thank you for your comments.
The issue  in my case is, in order to test my module, I need to interact with external protocol like Curve Finance, meaning that I need to run forking network with Foundry, which cause some issue. I have asked a help on community or your team, but seems no solution yet. That is why test cases are not attached. 
Alternatively, I have tested by hardhat java script separately as mentioned in initial comment.
https://github.com/kfujita0520/simplify-ubq-farming

@pavlovcik 
>the logic that I describe above is all implemented already in the instadapp smart contracts that

Thank you for your comment. Yes, but actually, I have made some improvement, for example, I have directly interacted with meta-pool when adding liquidity. Meaning that token like USDT, USDC, DAI, can directly be added into liquidity while previous contract required conversion from these tokens to 3CRV token once prior to adding liquidity.  

>this- is everything all right on your end?

I think i am all right on my end. If you still have some feedback, kindly please let me know. 
Thank you very much.

**Comment by 0x4007** [comment]
> The issue in my case is, in order to test my module, I need to interact with external protocol like Curve Finance, meaning that I need to run forking network with Foundry, which cause some issue. I have asked a help on community or your team, but seems no solution yet.

@0xcodercrane can you facilitate @kfujita0520 with this?

**Comment by cuddle-dev** [comment]
@kfujita0520 could you let us know which issue you're currently facing now? I've also checked your FoundryTest repo. it looked like not completed even in `setUp` function.  We can help hunters to resolve blockers as much as possible but you should complete the unit tests first. 

@ryzhak would you have a mind to help @kfujita0520  to write unit test in Foundry? We'll make sure you get paid for your help.

**Comment by ryzhak** [comment]
@0xcodercrane 

Yes, I can write tests for https://github.com/ubiquity/ubiquity-dollar/blob/4f0ac41db7b4435e0bbc59f0d242b23f22841c01/packages/contracts/src/dollar/DirectUBQFarmer.sol contract and create a PR to https://github.com/kfujita0520/ubiquity-dollar/tree/fujita repository

Should I start now or we are waiting for @kfujita0520 's confirmation?

**Comment by 0x4007** [comment]
> @0xcodercrane
> 
> Yes, I can write tests for https://github.com/ubiquity/ubiquity-dollar/blob/4f0ac41db7b4435e0bbc59f0d242b23f22841c01/packages/contracts/src/dollar/DirectUBQFarmer.sol contract and create a PR to https://github.com/kfujita0520/ubiquity-dollar/tree/fujita repository
> 
> Should I start now or we are waiting for @kfujita0520 's confirmation?

Allow a full 24 hours for @kfujita0520 to respond to your previous comment. If nothing then you should start.

**Comment by kfujita0520** [comment]
@0xcodercrane @pavlovcik 
Thank you for your comment. 
As spoken, this is my test case written as Foundry.
https://github.com/kfujita0520/Foundry_Test/blob/main/testProject/test/ForkTest.sol
I have wrote up setup function, but will cause the  EvmError: Revert error. 
If your team mate kindly support me to fix this issue, i can complete test case. 
@ryzhak  If you knew any clue to cause this error in my test case and let me know, it is really appreciated.
Meanwhile, as i said, i have done the test in hardhat javascript already for this contract.

**Comment by kfujita0520** [comment]
By the way, my test code of foundry is not completely garbage. If we could fix the said problem, it should be very close to complete.

**Comment by ryzhak** [comment]
@kfujita0520 

I will try to run the test case

**Comment by kfujita0520** [comment]
> @kfujita0520
> 
> I will try to run the test case

@ryzhak 
Thank you very much. please feel free to message me on telegram, if you have question on my foundry test script

**Comment by ryzhak** [comment]
PR with tests https://github.com/kfujita0520/ubiquity-dollar/pull/1

Basically I just mocked everything I could

@kfujita0520 you are free to change mocks to real implementation where you see fit

this is the best I can do for now because my understanding of the whole project is not deep enough

I also renamed contract filename from `DirectUBQFarmer.sol` to `DirectGovernanceFarmer.sol` because @pavlovcik updated contract name earlier in this PR

**Comment by 0x4007** [comment]
Thanks @ryzhak 

I suggest asking @0xcodercrane for their input when this is ready because they are deeply familiar with the codebase.

**Comment by kfujita0520** [comment]
@ryzhak 
Thank you very much for your update. It is very nice and clean unit test script. 
This is just question. Would you be able to write integration test script in Foundry in mainnet forking network? 
That is what I was struggled with and I suspect Foundry has some issue on it. (Some external contract cause issue in mainnet forking, but others are not and i don't find any clear condition nor solution even asking the community.)
I ended up with spending more time on this issue rather than contract development itself. So if you knew the answer, i would like to know. 
If not, that is no problem. Anyway, i have done integration test with hardhat java script already. 
Thank you again.

**Review by 0x4007 — APPROVED** [review]

**Comment by ryzhak** [comment]
@kfujita0520 

> Would you be able to write integration test script in Foundry in mainnet forking network?

Basically unmocked unit tests are pretty much the same as integration tests in a forked environment. Anyway I need a more deep understanding of the whole project to write such tests.

**Comment by kfujita0520** [comment]
@ryzhak 
Let me rephrase my question.
The following test case of mine will be failed with "EvmError: Revert error." at line 82 when running the mainnet forking environment. Though this is just normal approval process, it failed. (in hardhat it will never failed). That is what I was struggled to solve and still do not have the answer.
https://github.com/kfujita0520/Foundry_Test/blob/main/testProject/test/ForkTest.sol
If you could know any reason and could advise me, that is really appreciated. If not have immediate answer, that is no problem at all. 


> Basically unmocked unit tests are pretty much the same as integration tests in a forked environment. Anyway I need a more deep understanding of the whole project to write such tests.

**Comment by ryzhak** [comment]
> @ryzhak Let me rephrase my question. The following test case of mine will be failed with "EvmError: Revert error." at line 82 when running the mainnet forking environment. Though this is just normal approval process, it failed. (in hardhat it will never failed). That is what I was struggled to solve and still do not have the answer. https://github.com/kfujita0520/Foundry_Test/blob/main/testProject/test/ForkTest.sol If you could know any reason and could advise me, that is really appreciated. If not have immediate answer, that is no problem at all.
> 
> > Basically unmocked unit tests are pretty much the same as integration tests in a forked environment. Anyway I need a more deep understanding of the whole project to write such tests.

hard to say, seems everything is fine here https://github.com/kfujita0520/Foundry_Test/blob/main/testProject/test/ForkTest.sol#L82

**Comment by cuddle-dev** [comment]
Ok. Thanks for your contributions so far @kfujita0520  @ryzhak . To get this pull requested merged, we need to have unit tests working. We'll be fine whoever makes unit tests working. Just make sure all the unit tests included to this pull request. 

I guess @ryzhak can make a pull request to @kfujita0520 's `fujita` branch with the completed unit tests and @kfujita0520 merges it into `fujita`'s branch. If then, it will reflected here as well.

**Comment by ryzhak** [comment]
@0xcodercrane 

PR is already here https://github.com/kfujita0520/ubiquity-dollar/pull/1, it's up to @kfujita0520 to merge it

**Comment by kfujita0520** [comment]
@ryzhak 
Sorry for missing approval. I have just approved and merged. Thank you very much for your work.
@0xcodercrane 
Thank you very much for your confirmation. We have done.

**Comment by cuddle-dev** [comment]
I appreciate it for your guys cooperative work! Looks good to me. but lets keep it for a while until `development` branch got fixed. Due to my mistake, it has been broken. Lol

**Review by cuddle-dev — APPROVED** [review]

**Pull request merged**
